src/nlp/search.py

```diff
@@ -0,0 +1,366 @@
+        Add vectors to the index.
+        If the arrays are inside a certain column, you can specify it using the `column` argument.
+        """
+        if self.faiss_index is None:
+            size = len(vectors[0]) if column is None else len(vectors[0][column])
+            if self.string_factory is not None:
+                index = faiss.index_factory(size, self.string_factory)
+            else:
+                index = faiss.IndexFlatIP(size)
+            if self.device > -1:
+                self.faiss_res = faiss.StandardGpuResources()
+                self.faiss_index = faiss.index_cpu_to_gpu(self.faiss_res, self.device, index)
+            elif self.faiss_gpu_options is not None:
+                self.faiss_index = faiss.index_cpu_to_gpu_multiple(
+                    self.faiss_gpu_options.resource_vec,
+                    self.faiss_gpu_options.device_vec,
+                    index,
+                    self.faiss_gpu_options.cloner_options,
+                )
+            else:
+                self.faiss_index = index
+        for i in range(0, len(vectors), batch_size):
+            vecs = vectors[i : i + batch_size] if column is None else vectors[i : i + batch_size][column]
+            self.faiss_index.add(vecs)
+
+    def search(self, query: np.array, k=10):
+        """ Find the nearest examples indices to the query.
+
+            Args:
+                `query` (`np.array`): The query as a numpy array.
+                `k` (`int`): The number of examples to retrieve.
+
+            Ouput:
+                `scores` (`List[List[float]`): The retrieval scores of the retrieved examples.
+                `indices` (`List[List[int]]`): The indices of the retrieved examples.
+        """
+        assert len(query.shape) == 1 or (len(query.shape) == 2 and query.shape[0] == 1)
+        queries = query.reshape(1, -1)
+        if not queries.flags.c_contiguous:
+            queries = np.asarray(queries, order="C")
+        scores, indices = self.faiss_index.search(queries, k)
+        return scores[0], indices[0].astype(int)
+
+    def search_batch(self, queries: np.array, k=10):
+        """ Find the nearest examples indices to the queries.
+
+            Args:
+                `queries` (`np.array`): The queries as a numpy array.
+                `k` (`int`): The number of examples to retrieve.
+
+            Ouput:
+                `scores` (`List[List[float]`): The retrieval scores of the retrieved examples per query.
+                `indices` (`List[List[int]]`): The indices of the retrieved examples per query.
+        """
+        assert len(queries.shape) == 2
+        if not queries.flags.c_contiguous:
+            queries = np.asarray(queries, order="C")
+        scores, indices = self.faiss_index.search(queries, k)
+        return scores, indices.astype(int)
+
+
+class IndexableMixin:
+    """Add indexing features to `nlp.Dataset`"""
+
+    def __init__(self):
+        self._indexes: Dict[str, BaseIndex] = {}
```

Comment:
`Dict[str, Dict[str, BaseIndex]]` for `{column_name: {index_id: BaseIndex}}`?

with a default index_id like `faiss_index_1` i.e. `camel_case(BaseIndex.__class__.__name__) + f"_{len(self._indexes[column_name])}"` and no need to give the name of the index if you only have one?

We can also keep that for later and wait for people to ask for several indexes on the same columns in the issues (not sure it's really useful at the moment).

Reply:
Good catch on the typing :)
Not sure if we want to have farfetched names as it can be confusing when trying to query the right index.
I like the idea of not having to provide the name if we only have one index

Reply:
I think I'd rather allow the user to specify a `custom_index_name` (defaults to `column_name`) rather than having this `column_name>index_id` logic actually